drivers/timer/stm32_lptim_timer.c

```diff
@@ -0,0 +1,292 @@
+		LL_LPTIM_SetAutoReload(LPTIM1, COUNT_PER_TICK);
+	}
+
+	/* Start the LPTIM counter in continuous mode */
+	LL_LPTIM_StartCounter(LPTIM1, LL_LPTIM_OPERATING_MODE_CONTINUOUS);
+
+#ifdef CONFIG_DEBUG
+	/* stop LPTIM1 during DEBUG */
+	LL_DBGMCU_APB1_GRP1_FreezePeriph(LL_DBGMCU_APB1_GRP1_LPTIM1_STOP);
+#endif
+	return 0;
+}
+
+void z_clock_set_timeout(s32_t ticks, bool idle)
+{
+	/* new LPTIM1 AutoReload value to set (aligned on Kernel ticks) */
+	u32_t next_arr = 0;
+
+	ARG_UNUSED(idle);
+
+	/* ARROK bit validates previous write operation to ARR register */
+	while (LL_LPTIM_IsActiveFlag_ARROK(LPTIM1) == 0) {
+	}
+	LL_LPTIM_ClearFlag_ARROK(LPTIM1);
+
+#ifdef CONFIG_TICKLESS_KERNEL
+	if (ticks == K_FOREVER) {
+		/* disable LPTIM */
+		LL_APB1_GRP1_ForceReset(LL_APB1_GRP1_PERIPH_LPTIM1);
+		LL_APB1_GRP1_DisableClock(LL_APB1_GRP1_PERIPH_LPTIM1);
+		return;
+	}
+
+	/* passing ticks==1 means "announce the next tick",
+	 * ticks value of zero (or even negative) is legal and
+	 * treated identically: it simply indicates the kernel would like the
+	 * next tick announcement as soon as possible.
+	 */
+	if (ticks <= (s32_t)1) {
+		ticks = 1;
+	} else {
+		ticks = (ticks - 1);
+	}
+	/* maximise Tick to keep next_arr on 32bit values,
+	 * in anycase the ARR cannot exceed LPTIM_TIMEBASE
+	 */
+	if (ticks > (s32_t)0xFFFF) {
+		ticks = 0xFFFF;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	volatile u32_t last_cnt = 0;
+        /* read current counter value (cannot exceed 16bit) */
+
+	/* It should be noted that to read reliably the content
+	 * of the LPTIM_CNT register, two successive read accesses
+	 * must be performed and compared
+	 */
+	volatile u32_t cnt_value = LL_LPTIM_GetCounter(LPTIM1);
+	while (last_cnt < cnt_value) {
+		last_cnt = LL_LPTIM_GetCounter(LPTIM1);
+	}
+
+	/* calculate the next arr value (cannot exceed 16bit)
+	 * adjust the next ARR match value to align on Ticks
+	 * from the current counter value to first next Tick
+	 */
+	next_arr = (((last_cnt * CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+			/ LPTIM_CLOCK) + 1) * LPTIM_CLOCK
+			/ (CONFIG_SYS_CLOCK_TICKS_PER_SEC);
+	/* add count unit from the expected nb of Ticks */
+	next_arr = next_arr + ((u32_t)(ticks) * LPTIM_CLOCK)
+			/ CONFIG_SYS_CLOCK_TICKS_PER_SEC;
+
+	/* maximise to TIMEBASE */
+	if (next_arr > LPTIM_TIMEBASE) {
+		next_arr = LPTIM_TIMEBASE;
+	}
+	/* run timer and wait for the reload match */
+	LL_LPTIM_SetAutoReload(LPTIM1, next_arr);
+
+	k_spin_unlock(&lock, key);
+
+#endif /* CONFIG_TICKLESS_KERNEL */
+}
+
+u32_t z_clock_elapsed(void)
+{
+	volatile u32_t last_cnt = 0;
+
+	if (!IS_ENABLED(CONFIG_TICKLESS_KERNEL)) {
+		return 0;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	/* It should be noted that to read reliably the content
+	 * of the LPTIM_CNT register, two successive read accesses
+	 * must be performed and compared
+	 */
+	volatile u32_t cnt_value = LL_LPTIM_GetCounter(LPTIM1);
+	while (last_cnt < cnt_value) {
+		last_cnt = LL_LPTIM_GetCounter(LPTIM1);
```

Comment:
no
 this is the way the ref manual is requesting to read the LPTIM counter:
" It should be noted that to read reliably the content of the LPTIM_CNT register, two successive read accesses must be performed and compared"

Reply:
But you need the comparison to take care of case like I describe above. In the way it is now, there are rare cases where it would get stuck in the while loop for 2 seconds ... 
Here is an alternative proposal from mbed-os for consecutive read and while loop which will not get stuck for a complete wrap-around: 
https://github.com/ARMmbed/mbed-os/blob/master/targets/TARGET_STM/lp_ticker.c#L348

Reply:
reading the LPTIM counter register is now:
`	while (lp_time != LL_LPTIM_GetCounter(LPTIM1)) {
		lp_time = LL_LPTIM_GetCounter(LPTIM1);
	}
`